tools/cache/time-defined-cache_test.go

```diff
@@ -116,4 +117,65 @@ var _ = Describe("time defined cache", func() {
 		}).WithPolling(250 * time.Millisecond).WithTimeout(1 * time.Second)
 	})
 
+	It("should not allow two threads to get value in parallel", func() {
```

Comment:
I have some doubts about this test case. What is its purpose? Is it to test how exclusive locks work? Since it is a core golang functionality, I am not sure that we need to cover it. My main concern here is that the test is a bit hard to follow, and it seems to rely on timeouts. Are we sure it does not introduce flakiness under load?

Thoughts?

Reply:
> Is it to test how exclusive locks work?

No :)
It basically tests that if multiple threads call the `Get()` method together, one thread would re-calculate and all other threads would wait and finally get the same cached value. It's valuable to guarantee this behavior. I think it's especially important when we're dealing with a cache that involves locking, as it's very easy to change implementation in a way that this behavior no longer holds.

> My main concern here is that the test is a bit hard to follow

Yeah, I agree. When go routines are involved the code is always more complex =/
However if you can think of a way to simplify it I'll be happy to hear it.

> it seems to rely on timeouts. Are we sure it does not introduce flakiness under load?

Realistically, go routines calling a function is an almost immediate operation. I don't think there's any reason for it to be delayed. I've given a 3 second timeout to be on the safe side. I don't think it's practical for it to take any longer.

WDYT?

Reply:
Okay, maybe let's see if we can simplify the test a bit. Added a few more comments. Do they make sense?